modules/pallets/messaging-fees/src/impls.rs

```diff
@@ -0,0 +1,311 @@
+	sp_runtime::traits::*,
+};
+use sp_core::{sr25519, H256};
+
+use crate::{types::IncentivizedMessage, *};
+use ismp::{
+	events::Event as IsmpEvent,
+	messaging::{hash_request, Message},
+	router::{Request, RequestResponse, Response},
+};
+use pallet_hyperbridge::VersionedHostParams;
+use pallet_ismp::fee_handler::FeeHandler;
+use pallet_ismp_host_executive::HostParam::{EvmHostParam, SubstrateHostParam};
+
+impl<T: Config> Pallet<T>
+where
+	<T as frame_system::Config>::AccountId: From<[u8; 32]>,
+	u128: From<<T as pallet_ismp::Config>::Balance>,
+	T::AccountId: AsRef<[u8]>,
+{
+	fn verify_and_get_relayer(signer: &Vec<u8>, signed_data: &[u8; 32]) -> Option<T::AccountId>
+	where
+		T::AccountId: From<[u8; 32]>,
+	{
+		type Sr25519Signature = (sr25519::Public, sr25519::Signature);
+
+		if let Ok((pub_key, sig)) = Sr25519Signature::decode(&mut &signer[..]) {
+			if sp_io::crypto::sr25519_verify(&sig, signed_data, &pub_key) {
+				return Some(pub_key.0.into());
+			}
+		}
+		None
+	}
+
+	fn accumulate_protocol_fees(message: &Message, relayer_account: &T::AccountId) {
+		let requests = match message {
+			Message::Request(req) => req.requests.clone(),
+			_ => return,
+		};
+
+		let relayer_address = relayer_account.as_ref().to_vec();
+
+		for req in requests {
+			let commitment =
+				hash_request::<<T as pallet::Config>::IsmpHost>(&Request::Post(req.clone()));
+			let source_chain = &req.source;
+
+			if source_chain.is_evm() {
+				if let Some(per_byte_fee) = Self::get_per_byte_fee(&req.dest) {
+					let fee = per_byte_fee.saturating_mul(U256::from(req.body.len()));
+					if fee > U256::zero() {
+						pallet_ismp_relayer::Pallet::<T>::accumulate_fee_and_deposit_event(
+							source_chain.clone(),
+							relayer_address.clone(),
+							fee,
+						);
+					}
+				}
+			} else if source_chain.is_substrate() {
+				if let Some(fee) = CommitmentFees::<T>::take(&commitment) {
+					let fee_u256: U256 = u128::from(fee).into();
+
+					pallet_ismp_relayer::Pallet::<T>::accumulate_fee_and_deposit_event(
+						source_chain.clone(),
+						relayer_address.clone(),
+						fee_u256,
+					);
+				}
+			}
+		}
+	}
+
+	fn get_per_byte_fee(state_machine: &StateMachine) -> Option<U256> {
+		let host_params = pallet_ismp_host_executive::HostParams::<T>::get(state_machine)?;
+
+		let per_byte_fee = match host_params {
+			EvmHostParam(evm_host_param) => {
+				let fee = evm_host_param
+					.per_byte_fees
+					.iter()
+					.find(|fee| {
+						let hashed_chain_id = sp_io::hashing::keccak_256(&state_machine.encode());
+						fee.state_id == H256(hashed_chain_id)
+					})
+					.map(|fee| fee.per_byte_fee)
+					.unwrap_or(evm_host_param.default_per_byte_fee);
+				fee
+			},
+			SubstrateHostParam(VersionedHostParams::V1(substrate_params)) => {
+				let fee = substrate_params
+					.per_byte_fees
+					.get(state_machine)
+					.cloned()
+					.unwrap_or(substrate_params.default_per_byte_fee);
+				let fee_u128: u128 = fee.into();
+				U256::from(fee_u128)
```

Comment:
This will likely be in 10 decimals as is standard for substrate runtimes, but we're treating it like 18 decimals